examples/scripts/parser/kin.js

```diff
@@ -1,42 +1,62 @@
-      pointSize: 3,
-      sizeAttenuation: false,
-      lineWidth: 3
-    })
+    var shape = new NGL.Shape(master)
 
     kinemage.dotLists.forEach(function (dotList) {
-      if (!dotList.master.includes(master)) return
-      for (var i = 0, il = dotList.position.length / 3; i < il; ++i) {
-        var i3 = i * 3
-        var x = dotList.position[ i3 ]
-        var y = dotList.position[ i3 + 1 ]
-        var z = dotList.position[ i3 + 2 ]
-        var r = dotList.color[ i3 ]
-        var g = dotList.color[ i3 + 1 ]
-        var b = dotList.color[ i3 + 2 ]
-        shape.addPoint([ x, y, z ], [ r, g, b ], dotList.label[ i ])
-      }
+      if (!dotList.masterArray.includes(master)) return
+      var pointBuffer = new NGL.PointBuffer({
+        position: new Float32Array(dotList.positionArray),
+        color: new Float32Array(dotList.colorArray)
+      }, {
+        pointSize: 2,
+        sizeAttenuation: false,
+        useTexture: true
+      })
+      shape.addBuffer(pointBuffer)
     })
 
     kinemage.vectorLists.forEach(function (vectorList) {
-      if (!vectorList.master.includes(master)) return
-      for (var i = 0, il = vectorList.position1.length / 3; i < il; ++i) {
-        var i3 = i * 3
-        var x1 = vectorList.position1[ i3 ]
-        var y1 = vectorList.position1[ i3 + 1 ]
-        var z1 = vectorList.position1[ i3 + 2 ]
-        var x2 = vectorList.position2[ i3 ]
-        var y2 = vectorList.position2[ i3 + 1 ]
-        var z2 = vectorList.position2[ i3 + 2 ]
-        var r = vectorList.color1[ i3 ]
-        var g = vectorList.color1[ i3 + 1 ]
-        var b = vectorList.color1[ i3 + 2 ]
-        shape.addWideline([ x1, y1, z1 ], [ x2, y2, z2 ], [ r, g, b ], vectorList.label1[ i ])
-      }
+      if (!vectorList.masterArray.includes(master)) return
+      var widelineBuffer = new NGL.WidelineBuffer({
+        position1: new Float32Array(vectorList.position1Array),
+        position2: new Float32Array(vectorList.position2Array),
+        color: new Float32Array(vectorList.color1Array),
+        color2: new Float32Array(vectorList.color2Array)
+      }, {
+        linewidth: vectorList.width[0]
+      })
+      shape.addBuffer(widelineBuffer)
+    })
+
+    kinemage.ballLists.forEach(function (ballList) {
+      if (!ballList.masterArray.includes(master)) return
+      var sphereBuff = new NGL.SphereBuffer({
+        position: new Float32Array(ballList.positionArray),
+        radius: new Float32Array(ballList.radiusArray),
+        color: new Float32Array(ballList.colorArray)
+      }, {
+        useTexture: true
+      })
+      shape.addBuffer(sphereBuff)
+    })
+
+    var positionArray = []
+    var colorArray = []
+    kinemage.ribbonLists.forEach(function (ribbonList) {
+      if (!ribbonList.masterArray.includes(master)) return
+      positionArray.push(...ribbonList.positionArray)
+      colorArray.push(...ribbonList.colorArray)
     })
+    if (positionArray.length > 0) {
```

Comment:
For now it is ok to use a buffer here but we should add a method `shape.addTriangle` to the `Shape` class to get support for picking/hover labels here as well.

Reply:
Here is the commit (https://github.com/arose/ngl/commit/545cc9387777205acdb7cf71b8e3f430816ad421) where I added `shape.addPoint` and `shape.addWideline` if you want to give it a try.

Reply:
Sounds good, I'll give it a go.

Reply:
great, thanks!